operator/internal/controller/auto/update_controller.go

```diff
@@ -579,3 +432,127 @@ func outputsToSecret(owner *autov1alpha1.Update, outputs map[string]*agentpb.Out
 
 	return s, nil
 }
+
+// stream is an interface constraint for the response streams consumable by a
+// streamReader.
+type stream interface {
+	agentpb.UpStream | agentpb.DestroyStream | agentpb.PreviewStream | agentpb.RefreshStream
+}
+
+// streamReader reads an update stream until a result is received. The
+// reconcile session and underlying Update object are updated to reflect the
+// result, but no changes are written back to the API server.
+type streamReader[T stream] struct {
+	receiver grpc.ServerStreamingClient[T]
+	obj      *autov1alpha1.Update
+	u        *reconcileSession
+	l        logr.Logger
+}
+
+// Recv reads one message from the stream which may or may not contain a
+// result.
+func (s streamReader[T]) Recv() (getResulter[T], error) {
+	stream, err := s.receiver.Recv()
+	return getResulter[T]{stream}, err
+}
+
+// Result reads from the underlying stream until a Result is received or an
+// error is encountered. A non-nil error is returned if the stream is closed
+// prematurely or if a gRPC error is encountered. Importantly, if the
+// Automation API returns an Unknown error it is assumed that the operation
+// failed; in this case a failed Result is returned with nil error.
+func (s streamReader[T]) Result() (result, error) {
+	var res result
+
+	for {
+		stream, err := s.Recv()
+		if err == io.EOF {
+			break
+		}
+		if err != nil && status.Code(err) != codes.Unknown {
+			// Surface gRPC errors.
+			return nil, err
```

Comment:
What happens in this situation w.r.t. the status? Is the error assumed to be transient? 

You might get PermissionDenied for example, or a InvalidArgument, some may be retryable.

Reply:
Transient errors will be handled with "fast" backoff, and non-transient errors will get this "slow" backoff.

I expanded slow backoff from just the Unknown status to now also Unauthenticated, PermissionDenied, and InvalidArgument.

Context cancelation and deadline exceeded will be handled with fast backoff.

LMK if that sounds reasonable.